pkg/pillar/cmd/downloader/downloader.go

```diff
@@ -398,9 +398,18 @@ func handleModify(ctx *downloaderContext, key string,
 		status.Expired, status.Name)
 
 	// If RefCount from zero to non-zero and status has error
-	// or status is not downloaded then do install
-	if config.RefCount != 0 && (status.HasError() || status.State != types.DOWNLOADED) {
+	// or status is not downloaded or retry then do install
+	if config.RefCount != 0 && (status.HasError() ||
+		status.State != types.DOWNLOADED || config.RetryCount != 0) {
 		log.Functionf("handleModify installing %s", config.Name)
+		if config.RetryCount != 0 {
```

Comment:
Shouldn't we compare with the old value of RetryCount (taken from status)? And go inside this branch only if the new RetryCount is greater than the previous one. Otherwise, we will enter it on every new update of the config, not necessarily related to the retry attempts.

Reply:
The RetryCount from status and the RetryCount from config are not connected/synced. They count different events, the config.RetryCount measures how often we retry downloading a blob after image verification fails and the status.RetryCount measures how often we retry downloading an image when the downloading itself fails.

Your concern about retrying the download on every config update is valid. But, the only cases we change the DownloadConfig are: 1) `AddOrRefcountDownloaderConfig` when we are starting a new download, 2) `MaybeRemoveDownloaderConfig` when we are removing a download, and 3) `retryDownload`, the case we fix when we are retrying a download after image verification fails (the case we are interested in).

For `AddOrRefcountDownloaderConfig`, the RetryCount is, by default, 0. For `MaybeRemoveDownloaderConfig`, corner cases might result in unnecessary redownloads. I need to figure out how to address those.

Reply:
I added the condition to retry only when status.state == Downloaded. This should avoid retrying when we are add or removing DownloaderConfig.

Reply:
> The RetryCount from status and the RetryCount from config are not connected/synced. They count different events, the config.RetryCount measures how often we retry downloading a blob after image verification fails and the status.RetryCount measures how often we retry downloading an image when the downloading itself fails.

It sounds super confusing to me...
Usually, when we have `SomeEntityConfig.field` and `SomeEntityStatus.field` I expect them to describe what we intend (the config value) and what we really have now (the status value) for that `field`. But ok. In any case, we could have a stored state of the `RetryCount` (in its config semantics) in the status, so it's available in the `handleModify` func, and we can compare it with the old value.

Reply:
What you are saying makes sense to me. I will change the field name and add that field to the status.

Reply:
@OhmSpectator having a retryCount echo'ed back in the Status isn't useful and a potential source of different confusion and bugs.

If there was a retryCount in Status one could expect volumemgr to pick up the retryCount from Status and set the retryCount in the Config to that plus one. But that does not ensure that the retryCount in the Config increases (and that the Config get republished to to subscriber) since there could have been a +1 just before and Status has not yet been increased.

What we need here is not a retryCount (since volumemgr tracks the retries separate from the VerifyImageConfig, and the verifier doesn't look at the count) but a mechanism to have the verifier retry. 
Would it be more clear to introduce a timestamp and have volumemgr set VerifyImageConfig.LastRetry = time.Now()? That would also accomplish the job AFAICT

Reply:
I'm not sure I understand why echoing the counter in the Status can be confusing. I would expect the Status version of the counter to be set only from the Config, and the Config version to be set only by VolumeManager, regardless of the value in the Status.

The logic seems clear to me:

1. VolumeManager wants to redownload the blob → it increases the counter in the DownloaderConfig.
2. The Downloader gets the config, compares the counter value with the one it received previously, and then stores it into Status.

But okay, maybe in this case, the stored counter value shouldn't go into Status but rather into some private field in the Downloader context, as we don't expect it to be used by any other component...

Also, I'm not sure how we deal with VerifyImageConfig here. Until now, we’ve only handled DownloaderConfig.
Regarding the idea of using a timestamp in a *Config and making decisions based on it: how do we expect to compare this value in the target agent (the Downloader, I assume)? We set the timestamp in one component, publish the config with this new value to the target component, and then need to compare it to some other value to determine if it has been updated. But with a timestamp, we will always have Config.LastRetry != time.Now()...

Reply:
> But okay, maybe in this case, the stored counter value shouldn't go into Status but rather into some private field in the Downloader context, as we don't expect it to be used by any other component... Is it what you meant by echoing the counter in Status could be confusing?

Yes in the sense that the "requester" doesn't look at the Status count. But it doesn't hurt to have it there. Makes it easier to inspect it when debuggging things.

> Also, I'm not sure how we deal with VerifyImageConfig here. Until now, we’ve only handled DownloaderConfig. 

Sorry, I meant DownloaderConfig. My confusion.

> Regarding the idea of using a timestamp in a *Config and making decisions based on it: how do we expect to compare this value in the target agent (the Downloader, I assume)? We set the timestamp in one component, publish the config with this new value to the target component, and then need to compare it to some other value to determine if it has been updated. But with a timestamp, we will always have Config.LastRetry != time.Now()...

Yeah the downloader would still need to store the old value somewhere e.g., in DownloaderStatus. So it doesn't change anything fundamental.